sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/ImdsRetryStrategy.java

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.identity.implementation;
+
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.RequestRetryCondition;
+import com.azure.core.http.policy.RetryStrategy;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.function.Predicate;
+
+/**
+ * The retry strategy when sending a request for the IMDS endpoint.
+ */
+public class ImdsRetryStrategy implements RetryStrategy {
+    private static final int MAX_RETRIES = 5;
+    private final int maxRetries;
+    private final Duration baseDelay;
+    private final Predicate<RequestRetryCondition> shouldRetryCondition;
+
+    public ImdsRetryStrategy() {
+        this(MAX_RETRIES, Duration.ofMillis(800));
+    }
+    public ImdsRetryStrategy(int maxRetries) {
+        this(maxRetries, Duration.ofMillis(800));
+    }
+
+    public ImdsRetryStrategy(int maxRetries, Duration baseDelay) {
+        this.maxRetries = maxRetries;
+        this.baseDelay = baseDelay;
+        this.shouldRetryCondition = this::defaultShouldRetryCondition;
+    }
+
+    @Override
+    public int getMaxRetries() {
+        return maxRetries;
+    }
+
+    @Override
+    public Duration calculateRetryDelay(int retryAttempts) {
+        long delay = (long) (baseDelay.toMillis() * Math.pow(2, retryAttempts));
+        return Duration.ofMillis(delay);
+    }
+
+    @Override
+    public boolean shouldRetryCondition(RequestRetryCondition requestRetryCondition) {
+        return this.shouldRetryCondition.test(requestRetryCondition);
+    }
+
+    @Override
+    public boolean shouldRetry(HttpResponse httpResponse) {
```

Comment:
Does this retryStrategy + associated policy override the core policy or is it added to the pipeline? The reason I ask is because in JS at least we only need to account for 404, the rest are retried per the core retryPolicy

Reply:
Another thing I noticed in JS at least is that MSAL has its own retry behavior which we needed to disable with the `disableInternalRetries` flag otherwise we'd "double up" on retries - is that something that needs to happen / is already accounted for in Java?

Reply:
> Another thing I noticed in JS at least is that MSAL has its own retry behavior which we needed to disable with the `disableInternalRetries` flag otherwise we'd "double up" on retries - is that something that needs to happen / is already accounted for in Java?

It's different in .NET, also - we have to pass an extra argument to [WithHttpClientFactory](https://learn.microsoft.com/en-my/dotnet/api/microsoft.identity.client.baseabstractapplicationbuilder-1.withhttpclientfactory?view=msal-dotnet-latest#microsoft-identity-client-baseabstractapplicationbuilder-1-withhttpclientfactory(microsoft-identity-client-imsalhttpclientfactory-system-boolean))

Reply:
In Java, HTTP retries are controlled by the pipeline purely, as it overrides the http layer of msal.

Previously, In Java, we used to retry only on these specific error codes, so keeping it consistent.